apps/emqx_dashboard/src/emqx_dashboard_admin.erl

```diff
@@ -378,44 +451,153 @@ all_users() ->
-check(Username, Password) ->
+check(Username, Password, MfaToken) ->
     case lookup_user(Username) of
         [#?ADMIN{pwdhash = PwdHash} = User] ->
-            case verify_hash(Password, PwdHash) of
-                ok -> {ok, User};
-                error -> {error, <<"password_error">>}
-            end;
+            PasswordFn = fun() ->
+                case verify_hash(Password, PwdHash) of
+                    ok -> {ok, User};
+                    error -> {error, <<"password_error">>}
+                end
+            end,
+            MfaVerifyResult = verify_mfa_token(Username, MfaToken),
+            check_mfa(MfaVerifyResult, PasswordFn);
         [] ->
             {error, <<"username_not_found">>}
     end.
 
+check_mfa(ok, PasswordFn) ->
+    %% MFA token is OK
+    PasswordFn();
+check_mfa({error, MfaError}, PasswordFn) ->
+    %% MFA error
+    case emqx_dashboard_mfa:is_need_setup_error(MfaError) of
+        true ->
+            %% MFA needs setup, need to ensure password is OK.
+            %% Meaning: before MFA is setup, this user is still
+            %% vulnerable to brute-force attack (like no MFA).
+            case PasswordFn() of
+                {ok, _} ->
+                    %% password is OK
+                    %% Return the error which includes MFA setup
+                    %% secret or other context
+                    {error, MfaError};
+                {error, Reason} ->
+                    %% Return password error
+                    {error, Reason}
+            end;
+        false ->
+            %% Other errors, e.g.
+            %% - MFA is setup, but token is not provided
+            %% - Token is prvoided but stale
+            {error, MfaError}
+    end.
+
+verify_mfa_token(_Username, ?TRUSTED_MFA_TOKEN) ->
+    %% e.g. when handing change_pwd request which is authenticated
+    %% by bearer token
+    ok;
+verify_mfa_token(Username, MfaToken) ->
+    ok = maybe_init_mfa_state(Username),
+    do_verify_mfa_token(Username, MfaToken).
+
+do_verify_mfa_token(Username, MfaToken) when ?IS_NO_MFA_TOKEN(MfaToken) ->
+    %% MFA token is not provided
+    case get_mfa_enabled_state(Username) of
+        {ok, State} ->
+            %% Expected to receive MFA token, but not provided
+            {error, emqx_dashboard_mfa:make_token_missing_error(State)};
+        _ ->
+            %% No MFA state, proceed to password check
+            ok
+    end;
+do_verify_mfa_token(Username, MfaToken) ->
+    %% MFA token is provided
+    case get_mfa_enabled_state(Username) of
+        {ok, State} ->
+            %% MFA is configured, and a token is provided
+            case emqx_dashboard_mfa:verify(State, MfaToken) of
+                ok ->
+                    %% Token is ok, no state change
+                    ok;
+                {ok, NewState} ->
+                    %% Token is ok, state changed, should update db
+                    {ok, ok} = set_mfa_state(Username, NewState),
```

Comment:
Since this is a side effect that sets the MFA state, I think we should validate the password before doing this (currently it's the other way around).

Reply:
TOTP secret is only provided when password was once correct.
The second login attempt providing a valid TOTP code indicates that the authenticator app is set up.
Otherwise the user will get presented with the QR code again.

Reply:
Right, but it would add an extra layer of security if we require the password to be correct for any state altering side-effects to take place.  :thinking:

Reply:
I mean, it's multi-factor authn because there are multiple proofs given when authentication, e.g., password + TOTP.

Reply:
ok. fixed.